src/connector/src/sink/big_query.rs

```diff
@@ -82,23 +86,97 @@ pub struct BigQueryCommon {
     pub dataset: String,
     #[serde(rename = "bigquery.table")]
     pub table: String,
-    #[serde(rename = "bigquery.max_batch_rows", default = "default_max_batch_rows")]
-    #[serde_as(as = "DisplayFromStr")]
-    pub max_batch_rows: usize,
-    #[serde(rename = "bigquery.retry_times", default = "default_retry_times")]
-    #[serde_as(as = "DisplayFromStr")]
-    pub retry_times: usize,
     #[serde(default)] // default false
     #[serde_as(as = "DisplayFromStr")]
     pub auto_create: bool,
 }
 
-fn default_max_batch_rows() -> usize {
-    1024
+struct BigQueryFutureManager {
+    offset_queue: VecDeque<Option<TruncateOffset>>,
+    resp_stream: Pin<Box<dyn Stream<Item = Result<()>> + Send>>,
 }
+impl BigQueryFutureManager {
+    pub fn new(
+        max_future_num: usize,
+        resp_stream: impl Stream<Item = Result<()>> + Send + 'static,
+    ) -> Self {
+        let offset_queue = VecDeque::with_capacity(max_future_num);
+        Self {
+            offset_queue,
+            resp_stream: Box::pin(resp_stream),
+        }
+    }
+
+    pub fn add_offset(&mut self, offset: TruncateOffset, mut resp_num: usize) {
+        while resp_num > 1 {
+            self.offset_queue.push_back(None);
+            resp_num -= 1;
+        }
+        self.offset_queue.push_back(Some(offset));
+    }
 
-fn default_retry_times() -> usize {
-    5
+    pub async fn wait_next_offset(&mut self) -> Result<Option<TruncateOffset>> {
+        if let Some(Some(TruncateOffset::Barrier { .. })) = self.offset_queue.front() {
+            return Ok(self.offset_queue.pop_front().unwrap());
+        }
+        self.resp_stream
+            .next()
+            .await
+            .ok_or_else(|| SinkError::BigQuery(anyhow::anyhow!("end of stream")))??;
+        self.offset_queue.pop_front().ok_or_else(|| {
+            SinkError::BigQuery(anyhow::anyhow!(
+                "should have pending chunk offset when we receive new response"
+            ))
+        })
+    }
+}
+pub struct BigQueryLogSinker {
+    writer: BigQuerySinkWriter,
+    bigquery_future_manager: BigQueryFutureManager,
```

Comment:
Why not `future_manager`? we don't need to add so many prefixes in private fields/types. And I guess it may not be so suitable to call it "manager" here.

Also, do we really need a `BigQueryFutureManager` type here? It seems don't "manage" anything, we still directly access `bigquery_future_manager.offset_queue` below😂

Reply:
For async sink, we have a FutureManager to manage the asyn tasks and truncate our log store chunk after the async tasks are done. this BigQueryFutureManager  has the same tasks as FutureManager, just customized for bg sink